### ISSUE TITLE
Some thoughts on isomorphic rendering, dev-server and other stuff

### ISSUE THREAD

**Issue by timomeh** [Hashnode/mern-starter#58]
After fiddling around with MERN, I came across some thoughts I would like to share with you. A solid MERN kit/boilerplate is great, and making it more awesome would be … so much more awesome!

I'm aware that MERN is inspired by some well-known tutorials – but they aren't perfect. So let's try to make it better!

1. **Isolating webpack-dev-server and actual server**  
Since webpack-dev-server runs with express, and the "api-server" runs with express, it's very easy to combine both. But combining them causes webpack to initialize and rebuild everytime you need to restart the server (e.g. changes in server-controllers, which don't require a fresh webpack-build). Additionally a server-restart takes a few seconds to transpile the ES6 Syntax. All in all it takes too long for a satisfying experience.

2. **css/assets and isomorphic rendering**  
Currently MERN doesn't support importing css or assets (like images), because the server can't import/require these files. This is a problem I encountered myself several times and my only solution was to decide against isomorphic rendering. 😕 Maybe [webpack-isomorphic-tools](https://github.com/halt-hammerzeit/webpack-isomorphic-tools) could help, but I haven't found any time to try it.

3. **Directory structure**  
It's popular and used in some well-known tutorials to use `client`, `server`, `shared` and also to have the redux-things in `shared/redux/...`. Since MERN forces you to use redux (which is good, because redux is awesome), I don't think this structure necessary. E.g. digging through directories to open `shared/redux/constants/constants.js` is a bit of a hassle. Also, importing `../constants/constants` seems redundant.  
Directories help grouping related files together for a clear structure, but overdoing your directory structure causes the opposite: confusion and being annoyed. IMO a flatter directory structure would be more developer-friendly.  
Also, I'm sometimes thinking if differentiating between `client` and `shared` is really necessary. In the end, `shared` is the client-logic. You don't have any actual server-logic in the shared/client code. The server can "simply" use this code to pre-render the page (like fast-forwarding).
**tl;dr** A simpler directory structure would be less confusing.

So, yeah. I hope my feedback helps. Cheers. :beers:

**Comment by mayankchd** [comment]
Thanks a lot @timomeh for your feedback, we will look into it and make relevant changes!

**Comment by timomeh** [comment]
Sweet! Some of my proposals are big changes which need to be discussed, so I don't want to simply open a pull request. But if I can help, I'd be glad to help!

**Comment by sylvainlap** [comment]
About the 3rd point, it should be great to use a directory structure, with index.js files in it.

For example, in the `redux` directory, there should be a `actions` directory, a `constants` directory... with at the root of each one an `index.js`. So that a simple `import { action1, action2 } from '../actions'`, or `import * from '../constants'` would work.

**Comment by timomeh** [comment]
Is a dedicated directory for actions and constants really necessary? A `actions.js` and `constants.js` in no dedicated directory fulfills the same goal, without having a directory with a single file in it.

**Comment by sylvainlap** [comment]
Sure thing.

Maybe a directory could allow to have multiple files (for huge apps, where constants or actions should be splitted), and all these files could be exported into the index.js file.

But for a boilerplate, maybe only a file, and not a directory, is enough.

**Comment by timomeh** [comment]
This could also be a good way to "force users to structure their app". No one likes refactoring. 😉

For example, `constants.js` is just a list of constants. You won't have a few hundred lines of constants, so a single file is enough.  
But all actions in a single file would get very messy as your app grows. Separating actions into multiple files (`actions/PostAction.js`, ...) and a re-exporting `actions/index.js` is nice for both small and growing apps.

This could also be helpful for reducers. Maybe together with redux's `combineReducers` (or its more verbose VanillaJS-equivalent).

**Comment by prank7** [comment]
@timomeh @sylvainlap We definitely need to put actions in multiple files and use `combineReducers`, also I think some people like to keep constants for different parts of their app like say `admin` or `users` in different constant files thus dedicated directories for them doesn't sound like an awful idea to me. 😉

**Comment by isomoar** [comment]
What about 1st and 2nd points? I think that after implementation of this improvements + different directory structure MERN source code would be absolutely identical to [react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example). The only different is MongoDB present.

**Comment by somus** [comment]
We are going to work on v2 of MERN starting today. I've put up a discussion thread for the roadmap of v2 at #146. I've put things which were discussed here in that thread. We can continue the discussion there.
